internal/infrastructure/kubernetes/ratelimit/resource_provider.go

```diff
@@ -208,16 +217,16 @@ func (r *ResourceRender) Deployment() (*appsv1.Deployment, error) {
 		},
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace: r.Namespace,
-			Labels:    labels,
+			Labels:    rateLimitLabels(),
 		},
 		Spec: appsv1.DeploymentSpec{
 			Replicas: r.rateLimitDeployment.Replicas,
 			Strategy: *r.rateLimitDeployment.Strategy,
 			Selector: selector,
 			Template: corev1.PodTemplateSpec{
 				ObjectMeta: metav1.ObjectMeta{
-					Labels:      selector.MatchLabels,
-					Annotations: annotations,
+					Labels:      podLabels,
```

Comment:
these are immutable so updating the labels will result in a failed deployment
cc @sanposhiho

Reply:
To be accurate, .spec.selector is immutable, while .spec.template.labels isn't. But, we have to make sure the .spec.template.labels match .spec.selector, when updating it.

@oscarboher 
You can refer to the idea -
https://github.com/sanposhiho/gateway/blob/69bf8825422db0cb7b34097b8db5dfca19b8d477/internal/infrastructure/kubernetes/proxy/resource_provider.go#L195-L206
how we can make the selector stable while allowing patching labels.

So, basically, we should build the selector only with the labels that will never get changed. In other words, we shouldn't build the selector with the custom labels that users can change.

Also, you have to pay attention not to break the existing users who already have this deployment on their cluster.

Reply:
Currently the .spec.selector is a selector with the labels:
```
"app.kubernetes.io/name": InfraName,
"app.kubernetes.io/component":  "ratelimit",
"app.kubernetes.io/managed-by": "envoy-gateway",
```
So i understand this must remain the same to not break things because of the selector immutability.

Isn't it enough to ensure that the podLabels include at least those labels for the selector to work, as my PR does? It does look a bit funny using the copy function but it works.
What am I missing? Or the idea is just to have a method that can be used elsewhere to maintain some selectors?

Reply:
thanks @sanposhiho and @oscarboher, this change is safe since the `selector` isnt changing

Reply:
Oops, sorry, I thought `rateLimitLabels()` included `rateLimitDeployment.pod.labels` too. Yes, then nothing to worry here around the immutability.